### ISSUE TITLE
Please support using aws-lc-rs in place of ring

### ISSUE THREAD

**Issue by joshtriplett** [awslabs/aws-sdk-rust#1050]
### Describe the feature

I'd like to be able to use aws-lc-rs in place of ring, throughout my dependency tree. I'd love to be able to do so with aws-sdk-rust as well.

### Use Case

aws-lc-rs is faster. (EDIT: turns out it *doesn't* have a more compatible license, and still has code under the OpenSSL license.)

### Proposed Solution

I'd propose *either*:

1) Switching to aws-lc-rs as the default, or
2) Introducing feature flags to control which backend to use.

Switching to aws-lc-rs would be the simplest, and would be in line with the general aws-sdk-rust policy of not controlling functionality via feature flags.

Using a feature flag would allow people who prefer ring to continue using it.

I'd be happy to implement either solution.

### Other Information

https://github.com/awslabs/aws-sdk-rust/issues/966 requested support for the FIPS mode specifically, but that issue wasn't accepted because the FIPS mode requires Go to build.

However, the non-FIPS mode does *not* require Go. It does require cmake, but that seems somewhat less onerous. Nonetheless, if depending on cmake is a showstopper for this, I'd be happy to add a feature-flag-based solution, so that people who prefer to keep using ring can do so.

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### Community Note
* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue, please leave a comment

**Comment by jmklix** [comment]
We plan to add this at some point in the future, but I don't have a timeline for when this will be supported with this sdk

**Comment by joshtriplett** [comment]
@jmklix Would that timeline be affected if I wrote and submitted a patch implementing this?

(If so, which approach would you prefer? Switch over completely, or use a feature flag?)

**Comment by jdisanti** [comment]
Hey Josh,

We've talked about switching over to it as the default, but there has been some disagreement over that due to the requirement that cmake (and potentially even Go when FIPS is needed) be installed for it to compile. The ideal path forward would be to remove these requirements and make it default, but that's probably a significant amount of work on the aws-lc-rs side of things.

Our general stance is to not use cargo features for switching TLS implementations, so doing it that way is a no-go.

You should be able to work around this for now by swapping out the `HttpClient` implementation in config. There are examples in aws-config. Then you can turn off the `tls-rustls`/`rustls` features to remove the dependency on crate you have licensing issues with. If you want to make your solution available for others, you could also host a separate crate that does provides a `HttpClient` for aws-lc-rs until we get the default changed.

**Comment by joshtriplett** [comment]
@jdisanti
> We've talked about switching over to it as the default, but there has been some disagreement over that due to the requirement that cmake be installed for it to compile (and potentially even Go when FIPS is needed). The ideal path forward would be to remove these requirements and make it default, but that's probably a significant amount of work on the aws-lc-rs side of things.

I get that. It looks like https://github.com/aws/aws-lc-rs/issues/297 tracks that for aws-lc-rs, though I don't see a corresponding issue tracking that for aws-lc itself. That does seem like the ideal solution, if possible.

> You should be able to work around this for now by swapping out the HttpClient implementation in config. There are examples in aws-config. Then you can turn off the tls-rustls/rustls features to remove the dependency on crate you have licensing issues with.

I've found switching out `HttpClient` to be *remarkably* painful and error-prone, and I've had a number of issues attempting to do so in the past due to traveling the less well-trodden paths. I'm currently trying to *reduce* the degree of customization in my setup of the AWS SDK. That's why I'm trying to find a simpler solution than that.

Is that path (disabling those features) tested in CI these days?

**Comment by justsmth** [comment]
> We've talked about switching over to it as the default, but there has been some disagreement over that due to the requirement that cmake be installed for it to compile (and potentially even Go when FIPS is needed).

This limitation is being worked on: https://github.com/aws/aws-lc-rs/pull/317

**Comment by rcoh** [comment]
> I've found switching out HttpClient to be remarkably painful and error-prone, and I've had a number of issues attempting to do so in the past due to traveling the less well-trodden paths. I'm currently trying to reduce the degree of customization in my setup of the AWS SDK. That's why I'm trying to find a simpler solution than that.

> Is that path (disabling those features) tested in CI these days?

Yeah, there is a [CI test for it](https://github.com/smithy-lang/smithy-rs/tree/fd0034ec44cab3187d08766b7608b5142c694f63/aws/sdk/integration-tests/no-default-features). We also overhauled the way that HttpClients are configured, so I think that it is (mostly) a much improved experience. The one remaining issue is configuring a different HttpClient for credential providers.

**Comment by rcoh** [comment]
Another detail we'd probably want to iron out—I think we'd probably need to create some sort of `DoCrypto` trait of some sort to enable runtime switching without having to rewire everything

**Comment by yerke** [comment]
Now that https://github.com/aws/aws-lc-rs/pull/317 is merged and released, is it possible to reevaluate this issue please? Thanks.

**Comment by drauschenbach** [comment]
`aws-lc-sys` fails to build on my Arm systems and places like old Ubuntu. I'm forced to avoid it as much as possible when doing work that requires heterogeneity. So if this is done, I strongly suggest hiding `aws-lc-sys` behind a feature. Also, FWIW, a very large part of the Rust community wants pure-Rust ("turtles all the way down"), for a very long list of reasons.

**Comment by justsmth** [comment]
> aws-lc-sys fails to build on many Arm systems and places like old Ubuntu. 

Hello @drauschenbach!

Our team been been working on fixing [`aws-lc-rs`](https://github.com/aws/aws-lc-rs) build issues. We've made much progress recently, but there's still more work left to do.

When you have a chance, would you mind submitted an issue to our repo relating to which platforms you have trouble building for?  https://github.com/aws/aws-lc-rs/issues

Thanks for your feedback!

**Comment by aajtodd** [comment]
`aws-lc-rs` is now the default crypto provider, see this [discussion](https://github.com/awslabs/aws-sdk-rust/discussions/1257) and the [developer guide](https://docs.aws.amazon.com/sdk-for-rust/latest/dg/http.html) for more information
